tests/integration_tests/lightgbm_tuner_tests/test_optimize.py

```diff
@@ -732,6 +732,54 @@ def test_optuna_callback(self) -> None:
 
         assert callback_mock.call_count == 10
 
+    def test_tune_best_params_reproducibility(self) -> None:
+        import sklearn.datasets
+        from sklearn.model_selection import train_test_split
+
+        from optuna.samplers import TPESampler
+        import optuna.study as study
+
+        boston = sklearn.datasets.load_boston()
+        X_trainval, X_test, y_trainval, y_test = train_test_split(
+            boston.data, boston.target, random_state=0
+        )
+
+        train = lgb.Dataset(X_trainval, y_trainval)
+        valid = lgb.Dataset(X_test, y_test)
+        params = {"objective": "regression", "metric": "rmse", "random_seed": 0}
+
+        test_study_first_try = study.create_study(
+            direction="minimize", sampler=TPESampler(seed=10)
+        )
+
+        tuner_first_try = lgb.LightGBMTuner(
+            params,
+            train,
+            valid_sets=valid,
+            early_stopping_rounds=3,
+            study=test_study_first_try,
+            optuna_seed=10,
+        )
+        tuner_first_try.run()
+        best_params_first_try = tuner_first_try.best_params
+
+        test_study_second_try = study.create_study(
+            direction="minimize", sampler=TPESampler(seed=10)
+        )
+
+        tuner_second_try = lgb.LightGBMTuner(
+            params,
+            train,
+            valid_sets=valid,
+            early_stopping_rounds=3,
+            study=test_study_second_try,
+            optuna_seed=10,
+        )
+        tuner_second_try.run()
+        best_params_second_try = tuner_second_try.best_params
+
+        assert best_params_second_try == best_params_first_try
```

Comment:
In my environment, the best parameters are sometimes mismatched. As you can see the following log message, six runs out of ten failed. This may be due to the multiple parameters shows the same objective value. Do you think this is an environmental issue?

```console
$ pip install pytest-repeat
$ pytest --count=10 tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility
===================================================================================== short test summary info ======================================================================================
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[4-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[5-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[6-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[7-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[8-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
FAILED tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility[9-10] - AssertionError: assert {'bagging_fra...': False, ...} == {...
=================================================================================== 6 failed, 4 passed in 35.67s ===================================================================================
```
```

Reply:
@toshihikoyanase 
Thank you for your comment.
In fact, I tried same thing on my local pc and ended up with the following.. Could you consider to suggest how I could reproduce your result?

(base) C:\Users\xxx>pytest --count=10 D:\Git\Optuna_fix\optuna\tests\integration_tests\lightgbm_tuner_tests\test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility
====================================================================================== test session starts ======================================================================================
platform win32 -- Python 3.7.4, pytest-5.2.1, py-1.10.0, pluggy-0.13.1
rootdir: D:\Git\Optuna_fix\optuna
plugins: arraydiff-0.3, doctestplus-0.4.0, openfiles-0.4.0, remotedata-0.3.2, repeat-0.9.1
collected 10 items

D:\Git\Optuna_fix\optuna\tests\integration_tests\lightgbm_tuner_tests\test_optimize.py ..........                                                                                          [100%]

======================================================================================= warnings summary ========================================================================================
AppData\Roaming\Python\Python37\site-packages\win32\lib\pywintypes.py:2
  C:\Users\xxx\AppData\Roaming\Python\Python37\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp, sys, os

-- Docs: https://docs.pytest.org/en/latest/warnings.html
================================================================================ 10 passed, 1 warnings in 38.59s ================================================================================

(base) C:\Users\xxx>pytest --count=10 D:\Git\Optuna_fix\optuna\tests\integration_tests\lightgbm_tuner_tests\test_optimize.py::TestLightGBMTunerCV::test_tune_best_params_reproducibility
====================================================================================== test session starts ======================================================================================
platform win32 -- Python 3.7.4, pytest-5.2.1, py-1.10.0, pluggy-0.13.1
rootdir: D:\Git\Optuna_fix\optuna
plugins: arraydiff-0.3, doctestplus-0.4.0, openfiles-0.4.0, remotedata-0.3.2, repeat-0.9.1
collected 10 items

D:\Git\Optuna_fix\optuna\tests\integration_tests\lightgbm_tuner_tests\test_optimize.py ..........                                                                                          [100%]

======================================================================================= warnings summary ========================================================================================
AppData\Roaming\Python\Python37\site-packages\win32\lib\pywintypes.py:2
  C:\Users\xxx\AppData\Roaming\Python\Python37\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp, sys, os

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================================================== 10 passed, 1 warnings in 101.20s (0:01:41) ===========================================================================

Reply:
@toshihikoyanase 
As to the above (possibly) environment dependent issue, it would be appreciated if I could get any advice how we could move forward.

Reply:
@toshihikoyanase 
This is remainder since no update for two weeks. It would be appreciated if I could get any advice how we could move forward.

Reply:
FYI: In my local env, the test looks fine:

```zsh
❯ pytest --count=10 tests/integration_tests/lightgbm_tuner_tests/test_optimize.py::TestLightGBMTuner::test_tune_best_params_reproducibility
============================================================= test session starts =============================================================
platform darwin -- Python 3.8.8, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /Users/nzw/Documents/optuna, configfile: pyproject.toml
plugins: repeat-0.9.1
collected 10 items

tests/integration_tests/lightgbm_tuner_tests/test_optimize.py ..........                                                                [100%]

======================================================= 10 passed in 229.78s (0:03:49) ========================================================
>>> elapsed time 3m51s
```


## Dependencies and machine info

```
lightgbm                  3.1.1            py38h23ab428_0
```

OS: MacOS Big sur M1
OS version: 11.2.3 (20D91)

Reply:
@nzw0301 
Thank you for checking.
Could you consider to give advice how I could proceed to get approval of this PR?

Reply:
I'm sorry for the delayed response.

> As to the above (possibly) environment dependent issue

I agree. I investigated the cause, and found that the `study.best_trial` can return different trials if multiple trials have the same objective value. I think fixing `study.best_trial` is out-of-scope of this PR, so how about changing the criteria of reproducibility?

I suggest `LightGBMTuner` only ensures the same objective value in different runs, not the same parameter values.
What do you think of it?

Reply:
@toshihikoyanase 
Thank you for your feedback. I changed the test to compare not params but score and also changed the name of test to be consistent.